warehouse/warehouse.go

```diff
@@ -1066,6 +1079,228 @@ func processHandler(w http.ResponseWriter, r *http.Request) {
+		pkgLogger.Errorf("[WH]: pending-events:  Empty source id")
+		http.Error(w, "empty source id", http.StatusBadRequest)
+		return
+	}
+
+	pendingEvents := false
+
+	// check whether there are any pending staging files or uploads for the given source id
+	// get pending staging files
+	pendingStagingFileCount, err := getPendingStagingFileCount(sourceID)
+	if err != nil {
+		err := fmt.Errorf("Error getting pending staging file count : %v", err)
+		pkgLogger.Errorf("[WH]: %v", err)
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	// get pending uploads
+	pendingUploadCount, err := getPendingUploadCount(sourceID)
+	if err != nil {
+		err := fmt.Errorf("Error getting pending uploads : %v", err)
+		pkgLogger.Errorf("[WH]: %v", err)
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	// if there are any pending staging files or uploads, set pending events as true
+	if (pendingStagingFileCount + pendingUploadCount) > int64(0) {
+		pendingEvents = true
+	}
+
+	// trigger upload if there are pending events
+	if pendingEvents {
+		wh := make([]warehouseutils.WarehouseT, 0)
+
+		// get all wh destinations for given source id
+		destinationsMapLock.Lock()
+		for _, dst := range destinationsMap {
+			if dst.Source.ID == sourceID {
+				wh = append(wh, dst)
+			}
+		}
+		destinationsMapLock.Unlock()
+
+		// return error if no such destinations found
+		if len(wh) == 0 {
+			err := fmt.Errorf("No warehouse destinations found for source id '%s'", sourceID)
+			pkgLogger.Errorf("[WH]: %v", err)
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		for _, warehouse := range wh {
+			triggerUpload(warehouse)
+		}
+	}
+
+	// create and write response
+	res := warehouseutils.PendingEventsResponse{
+		PendingEvents: pendingEvents,
+	}
+
+	resBody, err := json.Marshal(res)
+	if err != nil {
+		err := fmt.Errorf("Failed to marshall pending events response : %v", err)
+		pkgLogger.Errorf("[WH]: %v", err)
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	w.Write(resBody)
+}
+
+func getPendingStagingFileCount(sourceID string) (fileCount int64, err error) {
+	var lastStagingFileID int64
+	sqlStatement := fmt.Sprintf(`SELECT end_staging_file_id FROM %[1]s WHERE %[1]s.source_id='%[2]s' ORDER BY %[1]s.id DESC`, warehouseutils.WarehouseUploadsTable, sourceID)
+
+	err = dbHandle.QueryRow(sqlStatement).Scan(&lastStagingFileID)
+	if err != nil && err != sql.ErrNoRows {
+		err = fmt.Errorf("Query: %s failed with Error : %w", sqlStatement, err)
+		return
+	}
+
+	sqlStatement = fmt.Sprintf(`SELECT COUNT(*)
+                                FROM %[1]s
+								WHERE %[1]s.id > %[2]v AND %[1]s.source_id='%[3]s'`,
+		warehouseutils.WarehouseStagingFilesTable, lastStagingFileID, sourceID)
+
+	err = dbHandle.QueryRow(sqlStatement).Scan(&fileCount)
+	if err != nil && err != sql.ErrNoRows {
+		err = fmt.Errorf("Query: %s failed with Error : %w", sqlStatement, err)
+		return
+	}
+
+	return
```

Comment:
explicitly return nil error ? since sql.ErrNoRows can also be considered not an error for our use case